### PR TITLE
Implement io::Seek for io::Empty

### LINKED ISSUE #78029: Implement Seek for io::Empty
You can create an empty io-reader with `std::io::empty()`, which returns an `std::io::Empty`. That `Empty` doesn't implement the `Seek` trait. However, I don't see any reason why it shouldn't.

### PR DESCRIPTION
**Pull request by oberien**
Fix #78029

### CODE REVIEW

**Comment by nagisa** [comment]
This makes sense to me, though exact behaviour when the seek would go past the 0-th position might need further discussion.

* `io::Cursor` currently returns `Err` if seek is past the end;
* `File` on UNIX will work just fine with seeking past the end and will properly set the position to that location;

And this introduces the 3rd behaviour where seeking past the end silently works, but does not update the position either.

**Comment by oberien** [comment]
The documentation of `Seek` states
> A seek beyond the end of a stream is allowed, but behavior is defined by the implementation. […]
> #### Errors
> Seeking to a negative offset is considered an error.

From that documentation my impression was that seeking beyond the end is totally fine. From my understanding that's also what's happening with `Cursor`:

https://github.com/rust-lang/rust/blob/dda2b5e3e260c14b868c494008af1c8981eaa5a8/library/std/src/io/cursor.rs#L210-L234

In that implementation, `Cursor::seek` only errors if the offset would underflow (go below 0 → seek to a negative position) or overflow (go above `u64::max_value()`). To me this seems similar to the behaviour you describe for `File` on unix.

I agree that it might make sense to mirror that behaviour of `Seek` to actually seek to that offset and return it as the new offset. However, right now `io::Empty` is a unit-struct and thus a zero-sized type. Adding a `pos: u64` member to it would increase its size to 8 bytes, which shouldn't be necessary for most uses. Thus I settled on the behaviour of this PR to always seek to position `0` without the need for any overhead.

File: library/std/src/io/util.rs
Comment (pickfire) [inline library/std/src/io/util.rs:131]:
```suggestion
}

```
It would be good to add an empty space here.

File: library/std/src/io/util.rs
Comment (pickfire) [inline library/std/src/io/util.rs:136]: How about `SeekFrom::Start(n)` being negative?

File: library/std/src/io/util.rs
Comment (oberien) [reply library/std/src/io/util.rs:131]: There wasn't an empty line between `impl Read for Empty` and `impl BufRead for Empty`. As I would put `impl Seek for Empty` among the same group of impls, I didn't put one there as well.

I don't mind either way though.

**Review by pickfire — APPROVED** [review]
Looks good to me.

**Comment by shepmaster** [comment]
Since traits are insta-stable, this probably needs a team sign-off.

r? @KodrAus

**Comment by oberien** [comment]
I rebased to the current master.

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

**Comment by oberien** [comment]
I rebased to the current master.

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

File: library/std/src/io/util.rs
Comment (jyn514) [inline library/std/src/io/util.rs:63]: This should be 1.50, 1.49 is currently in beta.
```suggestion
#[stable(feature = "empty_seek", since = "1.50.0")]
```

**Review by jyn514 — APPROVED** [review]
Would love to get eyes on this :) It would help quite a bit with https://github.com/algesten/ureq/pull/252.

File: library/std/src/io/util.rs
Comment (KodrAus) [inline library/std/src/io/util.rs:63]: Just dropping a comment here that this `since` field will need to be bumped if the FCP completes.

File: library/std/src/io/util.rs
Comment (KodrAus) [inline library/std/src/io/util.rs:67]: Should this also capture `SeekFrom::Start(n) if n.is_negative()`?

Comment (KodrAus) [reply library/std/src/io/util.rs:67]: It doesn't need to. `SeekFrom::Start` accepts a `u64`, not a `i64`.

**Comment by KodrAus** [comment]
@rfcbot fcp merge

This implements `Seek` for `io::Empty` such that:

- The stream length is reported as `0`.
- Seeking with a negative value returns an error.
- Seeking with any positive value will return `Ok`, but won't update the position.

File: library/std/src/io/util/tests.rs
Comment (pickfire) [inline library/std/src/io/util/tests.rs:35]:
```suggestion
    assert!(matches!(e.seek(SeekFrom::Start(0)), Ok(0)));
    assert!(matches!(e.seek(SeekFrom::Start(1)), Ok(0)));
```
Since we have these for the rest I think it would be good to add it too?

**Comment by m-ou-se** [comment]
The documentation of the Seek trait says:

> A seek beyond the end of a stream is allowed, but behavior is defined by the implementation. 

Does this sentence mean that all implementations always need to return `Ok` for any positive offset, and only the behaviour afterwards is implementation defined? Or is returning `Err` also a possible implementation-specific behaviour? (But then what does 'allowed' mean here?)

**Comment by sfackler** [comment]
I would interpret that as saying that the Seek impl shouldn't panic if you try to seek past the end, but could either return success or failure.

**Comment by m-ou-se** [comment]
Having `seek(Current(100));` succeed, but `seek(Current(-100));` fail afterwards doesn't seem great. I think either the first seek should fail, or both seek calls should succeed.

Some data points:

- Seeking in any direction on `/dev/null` on Linux always succeeds, even if the offset is negative. It always reports a position of 0. (So both calls would succeed. (Even in reverse order.))

- Seeking on a regular file or Cursor doesn't accept negative offsets, but it does allow `seek(Current(100)); seek(Current(-100));`, as it tracks the current position. (So both calls would succeed, when done in that order.)

However, since the only stable promise we're making here is that `io::Empty` implements `Seek`, without making any promises about the behaviour of negative or past-the-end seeks, merging this seems fine to me.

**Comment by sfackler** [comment]
Mirroring /dev/null's behavior seems pretty reasonable to me personally.

**Comment by oberien** [comment]
I agree that mirroring `/dev/null` could make sense here. I decided to error on negative seeks as the documentation of `Seek::seek` states
> ### Errors
>
> Seeking to a negative offset is considered an error.

I'm not sure if this is considered a guarantee, but there could be code that relies on `seek(Current(100)); seek(Current(-101))` to error. For example a function consuming a stream backwards could have the exit condition be an `Err` returned by the `seek`, which would result in an infinite loop if we were to mirror `/dev/null`.

File: library/std/src/io/util.rs
Comment (m-ou-se) [inline library/std/src/io/util.rs:72]:
```suggestion
    fn seek(&mut self, _pos: SeekFrom) -> io::Result<u64> {
        Ok(0)
    }
```

**Review by m-ou-se — COMMENTED** [review]
> I'm not sure if this is considered a guarantee, but there could be code that relies on `seek(Current(100)); seek(Current(-101))` to error. For example a function consuming a stream backwards could have the exit condition be an `Err` returned by the `seek`, which would result in an infinite loop if we were to mirror `/dev/null`.

That same code would break in the same way on `File::open("/dev/null")`.


> Mirroring /dev/null's behavior seems pretty reasonable to me personally.

> I don't mind either way.

Let's change that then before merging this.

**Comment by oberien** [comment]
I changed `io::Empty::seek` to follow `/dev/null` and squashed the commits back together into a single one.

**Comment by m-ou-se** [comment]
Great! Thank you!

@bors r+ rollup

**Pull request merged**
